imgpkg/002-recursive-bundles/README.md

```diff
@@ -0,0 +1,280 @@
+copy | exporting 4 images...
+copy | will export ghcr.io/k14s/design-docs/simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0
+copy | will export ghcr.io/k14s/design-docs/utilities@sha256:47ae428a887c41ba0aedf87d560eb305a8aa522ffb80ac1c96a37b16df038e0f
+copy | will export ghcr.io/k14s/design-docs/simple-app-bundle@sha256:d211dd700949154e429d28661d01c99d53a38af0d5275842ccbf0bf6dbef8ca4
+copy | will export ghcr.io/k14s/design-docs/simple-app-install-package@sha256:77c97e82306fb2a616da0a78796db039aa76e5bac1508954d40c0c10c073e035
+copy | exported 4 images
+copy | importing 4 images...
+copy | importing ghcr.io/k14s/design-docs/simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0 -> some-other.registry.io/recursive-bundle@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0...
+copy | importing ghcr.io/k14s/design-docs/utilities@sha256:ghcr.io/k14s/design-docs/utilities@sha256:47ae428a887c41ba0aedf87d560eb305a8aa522ffb80ac1c96a37b16df038e0f -> some-other.registry.io/recursive-bundle@sha256:ghcr.io/k14s/design-docs/utilities@sha256:47ae428a887c41ba0aedf87d560eb305a8aa522ffb80ac1c96a37b16df038e0f...
+copy | importing ghcr.io/k14s/design-docs/simple-app-bundle@sha256:d211dd700949154e429d28661d01c99d53a38af0d5275842ccbf0bf6dbef8ca4 -> some-other.registry.io/recursive-bundle@sha256:d211dd700949154e429d28661d01c99d53a38af0d5275842ccbf0bf6dbef8ca4...
+copy | importing ghcr.io/k14s/design-docs/simple-app-install-package@sha256:77c97e82306fb2a616da0a78796db039aa76e5bac1508954d40c0c10c073e035 -> some-other.registry.io/recursive-bundle@sha256:77c97e82306fb2a616da0a78796db039aa76e5bac1508954d40c0c10c073e035...
+copy | imported 4 images
+Succeeded
+```
+
+## Pull a bundle
+
+### Required changes
+
+#### Change folder structure
+
+When downloading a bundle that contains other bundles to disk using the `pull` command will work as currently, except for that it will download the nested bundles into a hidden folder called `.bundles`.
+
+##### Folder structure
+
+The structure of the output folder will be
+```
+$ tree -a recursive-bundle
+.
+├── .bundles
+│   ├── sha256-{SHA Of the First Nested Bundle}
+│   │   ├── .imgpkg
+│   │   │   ├── bundle.yml
+│   │   │   └── images.yml
+│   │   └── config2.yml
+│   └── sha256-{SHA Of the Second Nested Bundle}
+│   │   ├── .imgpkg
+│   │   │   ├── bundle.yml
+│   │   │   └── images.yml
+│       └── config1.yml
+├── .imgpkg
+│   ├── bundle.yml
+│   └── images.yml
+└── config.yml
+
+7 directories, 9 files
+```
+
+The folder name will be the sha256-{SHA} where SHA is the SHA256 of the bundle OCI Image. The mapping between the folder names and the origin images can be found in the `.imgpkg/images.yml` of the bundle that included this bundle
+
+###### Cyclic nesting
+
+To ensure that there is not cyclic nesting in the disk `imgpkg` will flatten the bundle structure to 1 level.
```

Comment:
❤️

Reply:
Does the loss of this structure potentially create a more confusing UX for users? Say we have a high level bundle that contains multiple offerings (sub bundles), each with their own apps (installable bundles). If I want to install one of those apps, it may be nice to know which product offering, aka parent bundle, the app was from, especially if there are app name collisions. As an alternative, we could validate on push and error if there are any cycles in the bundles, which would also give the potentially helpful guarantee  that any imgpkg created bundles don't have cycles. I am not strongly opinionated here, just a thought.

Reply:
Yes, this is definitely a hard part of the current implementation, and, it makes the UX a little hard. 
If we did prevent cyclic nesting we might be blocking an aggregator from using `imgpkg`. An example would be 2 applications, each on separate bundles, use the Rabbit bundle. If you want to create a bundle with these 2 applications you cannot, or you will have to break the individual application bundles and create your own bundle.
This can happen with Cloud Foundry or some other big system that contains multiple single-purpose applications that are bundled separately but you also want to provide a full installation.

To address some of these concerns I think we could eventually enhance this UX by allowing the bundle creators to provide an annotation that would tell `imgpkg` the name of the folder for a particular bundle in a later proposal. Do you think this would be helpful?

Note: Eventually if we were to provide such annotation we would have to ensure that there are no cyclical nesting.

Reply:
@timothysc is your comment about this ^
If so I was talking about a Rabbit Bundle that would contain Rabbit OCI Image + default configurations. 
Imagine this as the Bundle that the RabbitMQ team could provide for you to install in your cluster.

Reply:
Ah, yea, I guess if we preserved structure, since it is going on to a filesystem, each bundle would be limited to one parent, so we would have to write the bundle multiple times if it appeared multiple times, which seems less than ideal.

Reply:
So long as the lock file is clear I'm fine with flatting so long as users can nest to their hearts content provided there are 0 cycles. It needs to be a DAG.

Reply:
There will be no limit to the depth you can go with nested bundles. The only caveat is the more you add to a bundle the more time it will take to get everything.

The image lock file for each bundle will contain all the information that you would need to find the folder that will contain the bundle configuration you are looking for. Maybe not the location of the bundle but at least the information on how might you find it, like the SHA for the currently proposed placement in the disk.

Reply:
I believe the concerns in this thread were addressed in newer reviews of the proposal